ci/pingcap_tidb_operator_build_kind.groovy

```diff
@@ -27,22 +27,61 @@ def call(BUILD_BRANCH, CREDENTIALS_ID) {
 				def WORKSPACE = pwd()
 				dir("${PROJECT_DIR}"){
 					stage('build tidb-operator binary'){
-						checkout changelog: false, poll: false, scm: [$class: 'GitSCM', branches: [[name: "${BUILD_BRANCH}"]], doGenerateSubmoduleConfigurations: false, extensions: [], submoduleCfg: [], userRemoteConfigs: [[credentialsId: "${CREDENTIALS_ID}", refspec: '+refs/pull/*:refs/remotes/origin/pr/*', url: "${BUILD_URL}"]]]
+						if( BUILD_BRANCH ==~ /[a-z0-9]{40}/ ){
+							// checkout pull request
+							checkout changelog: false,
+							poll: false,
+							scm: [
+								$class: 'GitSCM',
+								branches: [[name: "${BUILD_BRANCH}"]],
+								doGenerateSubmoduleConfigurations: false,
+								extensions: [],
+								submoduleCfg: [],
+								userRemoteConfigs: [[
+									credentialsId: "${CREDENTIALS_ID}",
+									refspec: '+refs/pull/*:refs/remotes/origin/pr/*',
+									url: "${BUILD_URL}",
+								]]
+							]
+						} else {
+							// checkout branch, such as: master、release-1.0、release-1.1
+							checkout changelog: false,
+							poll: false,
+							scm: [
+								$class: 'GitSCM',
+								branches: [[name: "${BUILD_BRANCH}"]],
+								doGenerateSubmoduleConfigurations: false,
+								extensions: [],
+								submoduleCfg: [],
+								userRemoteConfigs: [[
+									credentialsId: "${CREDENTIALS_ID}",
+									url: "${BUILD_URL}",
+								]]
+							]
+						}
```

Comment:
fetching all branch/tags and PR refs is simpler, e.g. '+refs/heads/*:refs/remotes/origin/* +refs/pull/*:refs/remotes/origin/pr/*' (https://github.com/pingcap/tidb-operator/pull/1208/files)

Reply:
I tested this solution and found that it still doesn't work

Reply:
can you paste the failed e2e job link? I'd like to see how it works.

Reply:
I created a temporary job and run `checkout` only, it can checkout master branch. See the logs:

```
16:10:08  Cloning the remote Git repository
16:10:08  Cloning repository git@github.com:pingcap/tidb-operator.git
16:10:08   > git init /var/jenkins_home/workspace/operator_ghpr_e2e_test_kind_from_git2 # timeout=10
16:10:08  Fetching upstream changes from git@github.com:pingcap/tidb-operator.git
16:10:08   > git --version # timeout=10
16:10:08  using GIT_SSH to set credentials 
16:10:08   > git fetch --tags --progress git@github.com:pingcap/tidb-operator.git +refs/heads/*:refs/remotes/origin/*
16:10:30   > git config remote.origin.url git@github.com:pingcap/tidb-operator.git # timeout=10
16:10:30   > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
16:10:30   > git config remote.origin.url git@github.com:pingcap/tidb-operator.git # timeout=10
16:10:30  Fetching upstream changes from git@github.com:pingcap/tidb-operator.git
16:10:30  using GIT_SSH to set credentials 
16:10:30   > git fetch --tags --progress git@github.com:pingcap/tidb-operator.git +refs/heads/*:refs/remotes/origin/* +refs/pull/*:refs/remotes/origin/pr/*
16:10:38   > git rev-parse origin/master^{commit} # timeout=10
16:10:38  Checking out Revision ebc4486509e029ba81b1c3bc6b0de54962196eeb (origin/master)
16:10:38   > git config core.sparsecheckout # timeout=10
16:10:38   > git checkout -f ebc4486509e029ba81b1c3bc6b0de54962196eeb
16:10:38  Commit message: "Pin Terraform Google provider version to avoid use_ip_aliases error (#1206)"
```

https://internal.pingcap.net/idc-jenkins/job/operator_ghpr_e2e_test_kind_from_git2/1/console

Reply:
It simply runs a serial of git commands. It's easy to reproduce this locally.